rmw_zenoh_cpp/src/rmw_zenoh.cpp

```diff
@@ -2464,11 +2463,13 @@ rmw_send_request(
   // and any number.
   opts.consolidation = z_query_consolidation_latest();
   opts.value.payload = z_bytes_t{data_length, reinterpret_cast<const uint8_t *>(request_bytes)};
-  client_data->zn_closure_reply =
+  z_owned_closure_reply_t zn_closure_reply =
     z_closure(rmw_zenoh_cpp::client_data_handler, nullptr, client_data);
   z_get(
-    z_loan(context_impl->session), z_loan(
-      client_data->keyexpr), "", &client_data->zn_closure_reply, &opts);
+    z_loan(context_impl->session),
+    z_loan(client_data->keyexpr), "",
+    z_move(zn_closure_reply),
```

Comment:
Wondering if it makes sense for the `rmw_client` to own the callback. What happens when the callback does not get triggered yet (the service is busy), but then the client is destroyed? 

Would it be better to keep the original implementation but ensure we free the callback with `z_reply_null` when we call `rmw_client_fini`?

According to the [z_get documentation](https://zenoh-c.readthedocs.io/en/0.11.0/api.html#c.z_get):
> callback – The callback function that will be called on reception of replies for this query. Note that the reply parameter of the callback is passed by mutable reference, but will be dropped once your callback exits to help you avoid memory leaks. If you’d rather take ownership, please refer to the documentation of [z_reply_null()](https://zenoh-c.readthedocs.io/en/0.11.0/api.html#c.z_reply_null)

Reply:
Thanks for clarifying. I think it would be good to combine those changes in this PR as well.

> What you are talking about is the ownership of the reply.

Ah yeah sorry I got things mixed up but my original concern still stands on whether the callback gets freed if the client is destroyed before we receive a reply.

Reply:
> Ah yeah sorry I got things mixed up but my original concern still stands on whether the callback gets freed if the client is destroyed before we receive a reply.

This is a question for the Zenoh developers, but I believe not.  That's the reason for doing the `z_move`, so that zenoh-c/Zenoh proper is responsible for the lifetime of the closure.  I'll also point out that in my tests with valgrind, this didn't show a problem.